pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+package clockcache
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// CLOCK based approximate LRU storing mappings from strings to
+// int64s designed for concurrent usage.
+// It is sharded into 127 blocks, each of which can be
+// locked independently. Gets only require a read lock on the
+// individual block the element can be found in. When the cache is
+// not full, Insert only requires a write lock on the individual
+// block. Eviction locks blocks one at a time looking for a value
+// that's valid to evict/
+type Cache struct {
+	lock sync.RWMutex
+	// stores indexes into storage
+	elements map[interface{}]*Element
+	storage  []Element
+
+	// CLOCK sweep state, must have write lock
+	next int
+}
+
+type Element struct {
+	// The value stored with this element.
+	key   interface{}
+	Value interface{}
+
+	//CLOCK marker if this is recently used
+	used uint32
+
+	// pad Elements out to be cache aligned
+	_padding [24]byte
+}
+
+func WithMax(max uint64) *Cache {
+	if max < 1 {
+		panic("must have max greater than 0")
+	}
+	return &Cache{
+		elements: make(map[interface{}]*Element, max),
+		storage:  make([]Element, 0, max),
+	}
+}
+
+// Insert a key/value mapping into the cache if the key is not already present
+// returns the value present in the map, and true if it is newley inserted
+func (self *Cache) Insert(key interface{}, value interface{}) (canonicalValue interface{}, inserted bool) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	_, canonicalValue, inserted = self.insert(key, value)
+	return
+}
+
+// Insert a bactch of keys with their corresponding values.
+// This function will _overwrite_ the keys and values slices with their
+// canonical versions.
+func (self *Cache) InsertBatch(keys []interface{}, values []interface{}) {
+	if len(keys) != len(values) {
+		panic(fmt.Sprintf("keys and values are not the same len. %d keys, %d values", len(keys), len(values)))
+	}
+	values = values[:len(keys)]
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for idx := range keys {
+		keys[idx], values[idx], _ = self.insert(keys[idx], values[idx])
+	}
+	return
+}
+
+func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
+	elem, present := self.elements[key]
+	if present {
+		return elem.key, elem.Value, false
+	}
+
+	var insertLocation *Element
+	if len(self.storage) >= cap(self.storage) {
+		insertLocation = self.evict()
+		*insertLocation = Element{key: key, Value: value}
+	} else {
+		self.storage = append(self.storage, Element{key: key, Value: value})
+		insertLocation = &self.storage[len(self.storage)-1]
+	}
+
+	self.elements[key] = insertLocation
+	return key, value, true
+}
+
+func (self *Cache) evict() (insertPtr *Element) {
+	for {
+		insertLocation, evicted := self.tryEvict()
+		if evicted {
+			return insertLocation
+		}
+	}
+}
+
+func (self *Cache) tryEvict() (insertPtr *Element, evicted bool) {
```

Comment:
can't we rewrite cache.evict() as the following and avoid the try. The following should guarantee we always return
```
for {
      if self.next >= len(self.storage) {
		self.next = 0
      }
      elem := &self.storage[self.next]
      if elem.used != 0 {
         elem.used = 0
      }
      else {
	       delete(self.elements, elem.key)
               return elem
      }
      self.next += 1
}
```

Reply:
Think this would need to increment self.next before returning elem